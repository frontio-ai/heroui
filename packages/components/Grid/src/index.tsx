import React from "react";

import Grid from "./Grid";
import GridItem from "./GridItem";

// Example component showcasing different Grid configurations
const GridExample = () => {
  return (
    <Grid
      container
      columns={{
        "@initial": 1,
        "@sm": 2,
        "@md": 3,
      }}
      gap={{
        "@initial": "10px",
        "@sm": "20px",
        "@md": "30px",
      }}
    >
      <GridItem>Item 1</GridItem>
      <GridItem span={2}>Item 2 (span 2 columns)</GridItem>
      <GridItem>Item 3</GridItem>
      <GridItem>
        <Grid container columns={2} gap="10px">
          <GridItem>Nested Item 4.1</GridItem>
          <GridItem>Nested Item 4.2</GridItem>
        </Grid>
      </GridItem>
    </Grid>
  );
};

// Export the core Grid components and example
<<<<<<< HEAD
export {Grid, GridItem, GridExample};
=======
export {Grid, GridItem, GridExample}; // Include GridExample in exports
>>>>>>> 9207c6d06 (feat: grid component updates with fixes)
export type {GridProps} from "./Grid";
export type {GridItemProps} from "./GridItem";
