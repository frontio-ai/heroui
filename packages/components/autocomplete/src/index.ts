import type {ListboxItemProps, ListboxSectionProps} from "@vezham/listbox";
import type {MenuTriggerAction as BaseMenuTriggerAction} from "@react-types/combobox";

import {ListboxItem, ListboxSection} from "@vezham/listbox";

import Autocomplete from "./autocomplete";

// export types
export type {AutocompleteProps} from "./autocomplete";
export type {ListboxItemProps as AutocompleteItemProps};
export type {ListboxSectionProps as AutocompleteSectionProps};
export type MenuTriggerAction = BaseMenuTriggerAction | undefined;

// export hooks
export {useAutocomplete} from "./use-autocomplete";

// export components
export {Autocomplete, ListboxItem as AutocompleteItem, ListboxSection as AutocompleteSection};
