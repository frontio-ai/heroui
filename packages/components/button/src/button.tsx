import {Spinner} from "@nextui-org/spinner";
import {Ripple} from "@nextui-org/ripple";
import {forwardRef} from "@nextui-org/system";

import {UseButtonProps, useButton} from "./use-button";

export interface ButtonProps extends UseButtonProps {}

const Button = forwardRef<"button", ButtonProps>((props, ref) => {
  const {
    Component,
    domRef,
    children,
    styles,
    spinnerSize,
    spinner = <Spinner color="current" size={spinnerSize} />,
    spinnerPlacement,
    startContent,
    endContent,
    isLoading,
    disableRipple,
    getButtonProps,
    getRippleProps,
    isIconOnly,
  } = useButton({...props, ref});

  if (isIconOnly && isLoading) {
    return (
      <Component ref={domRef} className={styles} {...getButtonProps()}>
        {startContent}
        {isLoading && spinner}
        {endContent}
        {!disableRipple && <Ripple {...getRippleProps()} />}
      </Component>
    );
  }

  return (
    <Component ref={domRef} className={styles} {...getButtonProps()}>
      {startContent}
      {isLoading && spinnerPlacement === "start" && spinner}
      {children}
      {isLoading && spinnerPlacement === "end" && spinner}
      {endContent}
      {!disableRipple && <Ripple {...getRippleProps()} />}
    </Component>
  );
});

Button.displayName = "NextUI.Button";

export default Button;
