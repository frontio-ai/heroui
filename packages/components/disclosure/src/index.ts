import Disclosure from "./disclosure";

// export types
export type {DisclosureProps} from "./disclosure";

// export hooks
export {useDisclosure} from "./use-disclosure";

// export component
export {Disclosure};
