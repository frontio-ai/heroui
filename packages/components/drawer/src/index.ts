import {
  ModalHeader,
  ModalBody,
  ModalFooter,
  ModalContent,
  type ModalContentProps,
  type ModalHeaderProps,
  type ModalBodyProps,
  type ModalFooterProps,
} from "@nextui-org/modal";

import Drawer from "./drawer";

// export types
export type {DrawerProps} from "./drawer";
export type {
  ModalContentProps as DrawerContentProps,
  ModalHeaderProps as DrawerHeaderProps,
  ModalBodyProps as DrawerBodyProps,
  ModalFooterProps as DrawerFooterProps,
};

// export hooks
export {useDrawer} from "./use-drawer";

// export component
export {Drawer};

// export subcomponents
export {
  ModalHeader as DrawerHeader,
  ModalBody as DrawerBody,
  ModalFooter as DrawerFooter,
  ModalContent as DrawerContent,
};
