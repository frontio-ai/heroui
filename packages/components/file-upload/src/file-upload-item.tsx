import {Button} from "@nextui-org/button";
import {HTMLNextUIProps} from "@nextui-org/system";
import {CloseIcon} from "@nextui-org/shared-icons";

export interface FileUploadItemProps extends HTMLNextUIProps<"div"> {
  file: File;
  onFileRemove: (file: File) => void;
  isDisabled?: boolean;
}

const formatFileSize = (bytes: number) => {
  if (bytes === 0) return "0 Bytes";
  const k = 1024;
  const sizes = ["Bytes", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + " " + sizes[i];
};

const FileUploadItem: React.FC<FileUploadItemProps> = ({
  file,
  onFileRemove,
  isDisabled,
  ...otherProps
}) => {
  return (
    <div {...otherProps}>
      <Button
        aria-label={`Remove ${file.name}`}
        isDisabled={isDisabled}
        role="listitem"
        onClick={() => onFileRemove(file)}
      >
        <CloseIcon />
      </Button>
      <span>{file.name}</span>
      <span>{formatFileSize(file.size)}</span>
      <span>{file.type.split("/").at(1)?.toUpperCase()}</span>
    </div>
  );
};

FileUploadItem.displayName = "NextUI.FileUploadItem";

export default FileUploadItem;
