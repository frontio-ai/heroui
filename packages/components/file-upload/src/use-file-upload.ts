import type {FileUploadSlots, FileUploadVariantProps, SlotsToClasses} from "@nextui-org/theme";

import {HTMLNextUIProps, mapPropsVariants} from "@nextui-org/system";
import {fileUpload} from "@nextui-org/theme";
import {ReactRef, useDOMRef} from "@nextui-org/react-utils";
import {objectToDeps} from "@nextui-org/shared-utils";
import {ReactElement, useMemo} from "react";
import {ButtonProps} from "@nextui-org/button";

interface Props extends Omit<HTMLNextUIProps<"div">, "onChange"> {
  /**
   * Ref to the DOM node.
   */
  ref?: ReactRef<HTMLElement | null>;
  classNames?: SlotsToClasses<FileUploadSlots>;
  /**
   * A property to set initial files (which might be fetched) or to control files from outside of the component.
   */
  files?: File[];
  /**
   * Custom Browse Button.
   */
  browseButton?: ReactElement<ButtonProps>;
  /**
   * A different text for the browse button.
   */
  browseButtonText?: string;
  /**
   * Custom Add Button.
   */
  addButton?: ReactElement<ButtonProps>;
  /**
   * Custom Reset Button.
   */
  resetButton?: ReactElement<ButtonProps>;
  /**
   * Custom Upload Button.
   */
  uploadButton?: ReactElement<ButtonProps>;
  /**
   * Custom element to display buttons such as the browse button in customized design and order.
   */
  buttons?: (
    onBrowse: () => void,
    onAdd: () => void,
    onReset: () => void,
  ) => ReactElement<HTMLElement>;
  /**
   * If set to true, multiple files can be selected from the device.
   * @default false
   */
  multiple?: boolean;
  /**
   * Accept certain file format(s).
   */
  accept?: string;
  /**
   * Custom Element for an Upload File Item.
   */
  fileItemElement?: (file: File) => ReactElement<HTMLElement>;
  /**
   * Custom Element for topbar of the component.
   */
  topbar?: ReactElement<HTMLElement>;
  /**
   * Triggered when file(s) selected, added or removed.
   */
  onChange?: (files: File[]) => void;
}

export type UseFileUploadProps = Props & FileUploadVariantProps;

export function useFileUpload(originalProps: UseFileUploadProps) {
  const [props, variantProps] = mapPropsVariants(originalProps, fileUpload.variantKeys);

  const {ref, as, className, multiple = false, browseButtonText = "Browse", ...otherProps} = props;

  const Component = as || "div";

  const domRef = useDOMRef(ref);

  const styles = useMemo(
    () =>
      fileUpload({
        ...variantProps,
        className,
      }),
    [objectToDeps(variantProps), className],
  );

  return {
    Component,
    styles,
    domRef,
    className,
    multiple,
    browseButtonText,
    ...otherProps,
  };
}

export type UseFileUploadReturn = ReturnType<typeof useFileUpload>;
