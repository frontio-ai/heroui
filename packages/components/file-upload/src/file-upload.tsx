import {forwardRef} from "@nextui-org/system";
import {Button} from "@nextui-org/button";
import {cloneElement, useCallback, useEffect, useMemo, useRef, useState} from "react";
import {clsx} from "@nextui-org/shared-utils";

import {UseFileUploadProps, useFileUpload} from "./use-file-upload";
import FileUploadItem from "./file-upload-item";

export interface FileUploadProps extends UseFileUploadProps {}

const FileUpload = forwardRef<"div", FileUploadProps>((props, ref) => {
  const {
    Component,
    domRef,
    children,
    files: initialFiles,
    styles,
    className,
    classNames,
    multiple,
    browseButton,
    browseButtonText,
    addButton,
    resetButton,
    uploadButton,
    buttons,
    fileItemElement,
    topbar,
    onChange,
    ...otherProps
  } = useFileUpload({...props, ref});

  const inputFileRef = useRef<HTMLInputElement>(null);
  const singleInputFileRef = useRef<HTMLInputElement>(null);
  const [files, setFiles] = useState<File[]>(initialFiles ?? []);

  useEffect(() => {
    initialFiles && setFiles(initialFiles);
  }, [initialFiles]);

  const updateFiles = useCallback(
    (files: File[]) => {
      setFiles(files);
      onChange?.(files);
      // Setting input values to "" in order to ignore previously-selected file(s).
      // This will fix some bugs when "removing" and re-adding "the exact same" file(s) (e.g. removing foo.txt and adding foo.txt again).
      if (inputFileRef.current) inputFileRef.current.value = "";
      if (singleInputFileRef.current) singleInputFileRef.current.value = "";
    },
    [setFiles, onChange],
  );

  const topbarElement = useMemo(() => {
    if (topbar) {
      return cloneElement(topbar, {
        className: styles.topbar({class: clsx(classNames?.topbar, topbar?.props.className)}),
      });
    }
  }, [styles, classNames, topbar]);

  const items = useMemo(
    () =>
      files.map((file) => {
        if (!fileItemElement) {
          return (
            <FileUploadItem
              key={file.name}
              className={styles.item()}
              file={file}
              onFileRemove={(name) => {
                if (props.isDisabled) return;
                const newFiles = files.filter((file) => file.name !== name);

                updateFiles(newFiles);
              }}
            />
          );
        }

        const customFileElm = fileItemElement(file);

        return cloneElement(customFileElm, {
          key: file.name,
          className: styles.item({class: clsx(classNames?.item, customFileElm.props.className)}),
        });
      }),
    [styles, classNames, files, fileItemElement, updateFiles],
  );

  const onBrowse = useCallback(() => {
    if (props.isDisabled) return;
    inputFileRef.current?.click();
  }, [props.isDisabled]);

  const onAdd = useCallback(() => {
    if (props.isDisabled) return;
    singleInputFileRef.current?.click();
  }, [props.isDisabled]);

  const onReset = useCallback(() => {
    if (props.isDisabled) return;
    updateFiles([]);
  }, [props.isDisabled]);

  const browseButtonElement = useMemo(
    () =>
      browseButton ? (
        cloneElement(browseButton, {
          disabled: props.isDisabled,
          onClick: (ev) => {
            onBrowse();
            browseButton.props.onClick?.(ev);
          },
        })
      ) : (
        <Button disabled={props.isDisabled} onClick={() => onBrowse()}>
          {browseButtonText}
        </Button>
      ),
    [browseButton, browseButtonText, onBrowse, props.isDisabled],
  );

  const addButtonElement = useMemo(
    () =>
      addButton ? (
        cloneElement(addButton, {
          disabled: props.isDisabled,
          onClick: (ev) => {
            onAdd();
            addButton.props.onClick?.(ev);
          },
        })
      ) : (
        <Button color="secondary" disabled={props.isDisabled} onClick={() => onAdd()}>
          Add
        </Button>
      ),
    [addButton, onAdd],
  );

  const resetButtonElement = useMemo(
    () =>
      resetButton ? (
        cloneElement(resetButton, {
          disabled: props.isDisabled,
          onClick: (ev) => {
            onReset();
            resetButton.props.onClick?.(ev);
          },
        })
      ) : (
        <Button
          color="warning"
          disabled={props.isDisabled}
          onClick={() => {
            onReset();
          }}
        >
          Reset
        </Button>
      ),
    [resetButton, onReset],
  );

  const buttonsElement = useMemo(() => {
    if (!buttons) {
      const uploadButtonElement =
        uploadButton && cloneElement(uploadButton, {disabled: props.isDisabled});

      return (
        <div className={styles.buttons()}>
          {multiple && files.length !== 0 && addButtonElement}
          {files.length !== 0 && resetButtonElement}
          {browseButtonElement}
          {uploadButtonElement}
        </div>
      );
    }

    const customButtonsElement = buttons(onBrowse, onAdd, onReset);

    return cloneElement(customButtonsElement, {
      className: styles.buttons({
        class: clsx(classNames?.buttons, customButtonsElement.props.className),
      }),
    });
  }, [
    onBrowse,
    onAdd,
    onReset,
    styles,
    multiple,
    files,
    browseButtonElement,
    addButtonElement,
    resetButtonElement,
    uploadButton,
  ]);

  const baseStyles = styles.base({class: clsx(classNames?.base, className)});

  return (
    <Component ref={domRef} className={baseStyles} {...otherProps}>
      <input
        ref={inputFileRef}
        className="hidden"
        multiple={multiple}
        title="file upload"
        type="file"
        onChange={(ev) => {
          if (!ev.target.files?.length) return;
          const newFiles: File[] = [];

          for (let index = 0; index < ev.target.files.length; index++) {
            const file = ev.target.files.item(index);

            file && newFiles.push(file);
          }
          updateFiles(newFiles);
        }}
      />

      <input
        ref={singleInputFileRef}
        className="hidden"
        title="single file upload"
        type="file"
        onChange={(ev) => {
          const singleFile = ev.target.files?.item(0);

          if (!singleFile) return;
          if (files.find((file) => file.name === singleFile.name)) return;
          files.push(singleFile);
          updateFiles([...files]);
        }}
      />

      {topbarElement}

      <div className={styles.items()}>
        {children}
        {items}
      </div>

      {buttonsElement}
    </Component>
  );
});

FileUpload.displayName = "NextUI.FileUpload";

export default FileUpload;
