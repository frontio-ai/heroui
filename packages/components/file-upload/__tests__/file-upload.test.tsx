import * as React from "react";
import {render} from "@testing-library/react";
import {Button} from "@nextui-org/button";
import userEvent, {UserEvent} from "@testing-library/user-event";

import {FileUpload} from "../src";

describe("FileUpload", () => {
  let user: UserEvent;

  beforeEach(() => {
    user = userEvent.setup();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const wrapper = render(<FileUpload />);

    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("ref should be forwarded", () => {
    const ref = React.createRef<HTMLDivElement>();

    render(<FileUpload ref={ref} />);
    expect(ref.current).not.toBeNull();
  });

  it("should not browse when isDisabled", async () => {
    let renderResult = render(<FileUpload isDisabled />);
    let browseButton = renderResult.container.querySelector("button");

    expect(browseButton).toBeDisabled();

    renderResult = render(<FileUpload isDisabled browseButton={<Button>Browse files</Button>} />);
    browseButton = renderResult.container.querySelector("button");

    browseButton && (await user.click(browseButton));

    expect(browseButton).toBeDisabled();
  });

  it("buttons should be disabled when isDisabled", () => {
    const {container} = render(
      <FileUpload
        isDisabled
        multiple
        addButton={<Button>Add file</Button>}
        browseButton={<Button>Browse files</Button>}
        files={[new File([], "file1", {type: "jpg"})]}
        resetButton={<Button>Remove all files</Button>}
      />,
    );

    const buttons = container.getElementsByTagName("button");

    for (let index = 0; index < buttons.length; index++) {
      const btn = buttons.item(index);

      expect(btn).toBeDisabled();
    }
  });
});
