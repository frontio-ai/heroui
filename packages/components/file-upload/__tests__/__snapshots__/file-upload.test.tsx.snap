// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload should render correctly 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2"
  >
    <input
      class="hidden"
      title="file upload"
      type="file"
    />
    <input
      class="hidden"
      title="single file upload"
      type="file"
    />
    <div
      class="flex gap-3 p-3 w-full items-center overflow-hidden color-inherit subpixel-antialiased"
    >
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-default text-default-foreground data-[hover=true]:opacity-hover"
        type="button"
      >
        Browse
      </button>
    </div>
  </div>
</div>
`;
